Feature/autoconfig search handshake

tomereli opened the pull request:
Currently, prplMesh agent and controller always send VS TLVs to some messages even though it is possible the other side will not process these TLVs. Moreover, some testbed certified units don't cope well with fragmented CMDUs (R1 certification has no test for fragmentation/reassembly), so when adding a VS TLV big enough, the CMDU gets fragmented in the transport and then ignored by these units, causing autoconfiguration to fail.

It is possible to avoid this altogether by adding a handshake during the autoconfig search-reply phase - the prplMesh agent will add a VS TLV to the autoconfig search, and the controller will mark this agent as prplMesh if it contains this VS TLV and will answer it with an autoconfig response with a similar VS TLV. The agent will then understand that the controller is also prplMesh, so VS TLVs and VS CMDUs can be used.

This PR adds the basic handshake using an empty handshake request and reply messages attached to the autoconfig search and response messages.
By doing this, we mark the prplmesh controller as prplmesh or not prplmesh, and can use it to decide if to send VS TLVs and VS messages later on in the agent.

In addition, update the agent's node in the controller DB so that in a future PR we can use it to refrain from sending VS messages to non-prplmesh agents.

Fixes #675 
Fixes #425 
Fixes #423 
Fixes #424 
Fixes #921 

Passing 4.12.1 - https://gitlab.com/prpl-foundation/prplMesh/pipelines/124523357
Passing 4.12.2 - https://gitlab.com/prpl-foundation/prplMesh/-/jobs/464474890

Commit:
btlvf: add cACTION_CONTROL_SLAVE_HANDSHAKE_REQUEST

Preparative commit.
Add cACTION_CONTROL_SLAVE_HANDSHAKE_REQUEST which will be added to the
autoconfig search request to mark the agent as prplMesh.

Also add ACTION_CONTROL_SLAVE_HANDSHAKE_REQUEST to eActionOp_CONTROL
enum.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: update autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: autoconfig search add handshake vs tlv

Preparative commit.
In order to identify the agent as prplMesh, add a vendor specific TLV to
the autoconfig search message, so that if the controller is also
prplMesh, it will respond with an autoconfig search response with a
vendor specific TLV with a handshake response.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: db: add is_prplmesh and set_prplmesh

Preparative commit.
Add is_prplmesh and set_prplmesh APIs to the controller database which
will be used to mark prplmesh agents.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: add cACTION_CONTROL_SLAVE_HANDSHAKE_RESPONSE

Preparative commit.
Add cACTION_CONTROL_SLAVE_HANDSHAKE_RESPONSE which will be sent by the
controller as part of an autoconfiguration search response to mark
itself as prplmesh controller.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: update autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: add handshake response

Preparative commit.
Add handshake response VS TLV to autoconfig search response, which marks
the controller as prplMesh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: mark prplmesh controller

Preparative commit.
Mark prplmesh controller in backhaul manager and propagate it to the
slaves (radio agents).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: replace no_vendor_specific with prplmesh_controller

Now that we mark the controller as prplmesh or not, we can replace the
check which uses the flag no_vendor_specific, so do that.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide: remove no_vendor_specific

Its not used anymore.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalii-komisarenko commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -30,18 +30,20 @@ eActionOp_CONTROL:
-  ACTION_CONTROL_ARP_QUERY_REQUEST: 10
-  ACTION_CONTROL_ARP_QUERY_RESPONSE: 11
-  ACTION_CONTROL_PLATFORM_OPERATIONAL_NOTIFICATION: 12
+  ACTION_CONTROL_SLAVE_HANDSHAKE_REQUEST: 1
```
It breaks backward compartibility.
Are there on the market any older devices?

tomereli replied:
prplMesh is not deployed it, it is in development, so no.

vitalii-komisarenko commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -3599,3 +3599,16 @@ void db::clear_vap_list()
+bool db::is_prplmesh(const std::string &mac)
+{
+    auto node = get_node(mac);
+    return node ? node.get()->is_prplmesh : false;
```
Maybe I should have added a debug output here to signal that the node is not in the DB yet and we read `is_prplmesh` too early.

tomereli replied:
Fixed.

Commit:
agent: don't send VS messages for non prplmesh controllers

When the controller is not prplmesh, there is no point in sending vendor
specific messages which are proprietary prplmesh messages used for
communication between prplmesh controller and agents outside of the
easymesh specification, so stop doing that now.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

rmelotte commented:
Also fixes the issue seen in https://github.com/prplfoundation/prplMesh/issues/921, so I linked it.

rmelotte reviewed (approved):
:rocket: 
I added the don't merge tag just in case you want to squash the commits updating the generated files.

tomereli commented:
> 🚀
> I added the don't merge tag just in case you want to squash the commits updating the generated files.

Good thing you did - I'd like @arnout to also review this before we merge.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -3600,3 +3600,20 @@ void db::clear_vap_list()
 
 const std::shared_ptr<db::vaps_list_t> db::get_vap_list() { return m_vap_list; }
+
+bool db::is_prplmesh(const std::string &mac)
```
I *don't* want new interfaces that use strings for MAC addresses.

tomereli replied:
Ok.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -3600,3 +3600,20 @@ void db::clear_vap_list()
+void db::set_prplmesh(const std::string &mac)
+{
+    if (!get_node(mac)) {
+        add_node(network_utils::mac_from_string(mac));
```
I don't think a new node should be added in this case. The type of the node is going to be wrong. When we implement #524 we will need to know which type of node to create, so this isn't even possible.

tomereli replied:
I think we create the node only in during the autoconfig M1, so the node will always be created first when we get the autoconfig search request arrives, and if we don't create it here it we won't mark it as prplmesh when we mean to.
@vitalii-komisarenko please confirm.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2197,6 +2198,16 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
+                   << src_mac;
+        prplmesh_controller = true;
+    } else {
+        LOG(DEBUG) << "Not prplMesh controller " << src_mac;
```
Since we get multiple responses in the backhaul manager, perhaps we should log an error message if we already had prplmesh_controller == true. And we should probably reset it to false as well.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -380,6 +380,7 @@ bool backhaul_manager::finalize_slaves_connect_state(bool fConnected,
         network_utils::iface_info iface_info;
         bool backhaul_manager_exist = false;
 
+        notification->params().prplmesh_controller = pSocket->prplmesh_controller;
```
I don't think there's a need to save this flag per socket. We can just use `this->prplmesh_controller` here.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -108,6 +108,7 @@ class backhaul_manager : public btl::transport_socket_thread {
         bool sta_iface_filter_low        = false;
         bool slave_is_backhaul_manager   = false;
         bool controller_discovered       = false;
+        bool prplmesh_controller         = false;
```
So I think this member is unnecessary.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -228,6 +229,7 @@ class backhaul_manager : public btl::transport_socket_thread {
     bool pending_enable = false;
 
     std::string controller_bridge_mac;
+    bool prplmesh_controller = false;
```
Nitpick: I prefer names of booleans to be clearly boolean, so `is_prplmesh_controller` or `controller_is_prplmesh`.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -579,7 +579,7 @@ sBackhaulParams:
   backhaul_is_wireless: uint8_t 
   backhaul_iface_type: uint8_t 
   is_backhaul_manager: uint8_t 
-  gw_manage_type: uint8_t 
```
Removal of this unused field should be made explicit in the commit message.

tomereli replied:
Oops I forgot to update the commit message, mergify beat me to it

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1942,7 +1942,7 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
 
-        if (config.no_vendor_specific) {
+        if (!backhaul_params.prplmesh_controller) {
             LOG(DEBUG) << "non-Intel, not adding ClientAssociationEvent VS TLV";
```
Maybe also change this to "non-prplmesh" then?

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3737,6 +3737,9 @@ bool slave_thread::send_cmdu_to_controller(ieee1905_1::CmduMessageTx &cmdu_tx)
     }
 
     if (cmdu_tx.getMessageType() == ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE) {
+        if (!backhaul_params.prplmesh_controller) {
```
Ideally, we should never even get here... So maybe throw a warning if we do?

tomereli replied:
NACK - I did this here instead of not sending the messages and checking if the controller is prplMesh everywhere else where we send the VS message.

arnout commented on tests/test_gw_repeater.sh:
```diff
@@ -77,12 +77,10 @@ main() {
+            ${rootdir}/tools/docker/run.sh -u ${UNIQUE_ID} ${VERBOSE_OPT} ${FORCE_OPT} start-agent -d -n ${repeater} -m aa:bb:cc:$index$index -- "$@"
             index=$((index+1))
-            no_vendor=
         done
```
So, this means we no longer have a test of an agent that acts as a non-prplmesh agent. Is that acceptable?

Maybe it's better to keep all this and just force the controller_is_prplmesh option to false if the no_vendor_specific option is set.

tomereli replied:
I think it is acceptable since we are testing agent certification on a nightly basis with non-prplmesh controllers and controller certification with non-prplmesh agents.

arnout reviewed (approved):
Nothing major so approving. Well, I consider the string MAC address a major issue but I assume you can fix it up :-)

There's a typo in one of the commit messages: resonse instead of response.
